Feature/ssr support (#810)
City-of-Helsinki/helsinki-design-system: site/src/docs/foundation/guidelines/server-side-rendering.mdx

- If you add more HDS components later, you do not need to remember to add their styles separately.
- If you happen to remove HDS components, you do not need to remember to remove their respective styles.

Let's go through a simple example of the usage of the tool:

```js
import { getCriticalHdsRules, hdsStyles } from "hds-react";

const criticalHdsRules = await getCriticalHdsRules(bodyHTML, hdsStyles);

const finalHtml =
<html>
  <head>
    ...
    <style data-used-styles dangerouslySetInnerHTML={{ __html: criticalHdsRules }} />
  </head>
  <body>
    ...
  </body>
<html>
```

First, we import the tool that is named `getCriticalHdsRules`. To be able to call it, we need two things:

1. The initially to be rendered Html body code as a string.
2. hdsStyles - a variable holding all the styles of the HDS react components.

Calling it returns a string containing the critical CSS styles. The rest is easy, we set those styles into
a style tag, that we will then inject into the finally rendered Html document. It can be wise to cache the result of
`getCriticalHdsRules` based on the function parameters, to improve performance.

See below for more complete examples:

> <Accordion

Reviewer: Just a minor thing: add prop `language="en"` to these Accordions so the close button text is not in Finnish which is the default.

Author: - [ ] Change accordion close button language to english

---
slug: "/foundation/guidelines/server-side-rendering"
title: "Server-side rendering"
navTitle: 'Server-side rendering'
---
import { Accordion, Notification } from 'hds-react';
import LeadParagraph from '../../../components/LeadParagraph';

# Server-side rendering

<LeadParagraph>
  The server-side rendering support of the Helsinki Design System makes the user's landing experience to pages smooth without
  flashes of unstyled content.
</LeadParagraph>

## What is server-side rendering?

> Server-side rendering (SSR) is the process where the rendering of html pages is done on the server-side. The fully rendered html

Reviewer 2: html -> HTML

Author: - [ ] html -> HTML

> document is then sent to the browser to be viewed. The rendering of the html can happen at build time (Static site generation

Reviewer 2: Maybe remove the "to be viewed" because the browser also renders it but it might confuse more in this context

Author: - [ ] Remove 'to be viewed'

> or pre-rendering), or during an HTTP request. The alternative to SSR is client-side rendering (CSR), where rendering of the
> html happens in the browser.

Reviewer 2: ```suggestion
The alternative to SSR is client-side rendering (CSR), where most of the HTML content is composed and rendered in the browser using Javascript"
```

Author: - [ ] Improve sentence

## How does HDS support server-side rendering?

For HDS components to work with server-side rendering, in server you need to inject the critical CSS styles

> of the HDS components that are being used into the initially rendered HTML. For hds-core, you must include the styles from

Reviewer: "rendering, in **the** server" . Although a better structure might be to put the server mention in to the end (SPOTPA). Maybe something like this: "--rendering, you need to inject the critical CSS styles of the HDS components that are being used into the initially rendered HTML on the server's side."

Author: - [ ] rendering, in server you need to inject the critical CSS styles of the HDS components that are being used into the initially rendered HTML. -> rendering, you need to inject the critical CSS styles of the HDS components that are being used into the initially rendered HTML on the server's side.

> provided CSS files yourself. For hds-react, HDS provides multiple options, that we will cover next.

Reviewer: "from **the** provided CSS files yourself. For hds-react HDS provides multiple options, which we will cover next."
Use which with commas and in sentences which are disposable, i.e. stand-alone sentences that you can remove and only provide additional detail.

Author: - [ ] Add <b>the</b>

<Notification label="Critical CSS" className="siteNotification">

> Critical CSS is the CSS applied to above-the-fold elements. It is the CSS of the immediately visible content in the browser

Reviewer: A lot of repetition of the word CSS. Maybe something like this is a bit better: "Critical CSS is applied to the above-the-fold elements. It provides the styles for the immediately visible content in the browser."

Author: - [ ] Refactor to: Critical CSS is applied to the above-the-fold elements. It provides the styles for the immediately visible content in the browser...

> <Notification label="Critical CSS" className="siteNotification">
>   Critical CSS is the CSS applied to above-the-fold elements. It is the CSS responsible for content immediately
>   visible when a user opens your website. Pages can be long with scrollable content, that is not visible immediately.
>   There's no need to include the styles of elements that are not visible. There is one exception though, and that is

Reviewer 2: ```suggestion
Critical CSS is the CSS applied to above-the-fold elements. It is the CSS of the immediately visible content in the browser viewport when a user opens your website. Critical CSS does not usually have the styles of the elements that are in the scrollable content outside of the browser viewport. There is one exception though...
```

>   if you have anchor links on the page. When user opens website from an url that includes an anchor link, browser

Reviewer 2: **the** browser automatically

Author: - [ ] Add <b>the</b>

> When user opens website from an url that includes an anchor link, the browser automatically scrolls the page so that the link

Reviewer: url => URL

Author: - [ ] url -> URL

  automatically scrolls the page so that the link will be visible.
</Notification>

### Option 1: getCriticalHdsRules tool to get critical styles (recommended)

> HDS React components use CSS-in-JS in such a way, that the styles are by default injected to head tag in browser. If you

Reviewer 2: ~~to~~ **into the** head in **the** browser

Author: - [ ] Fix articles

> HDS React components use CSS-in-JS in such a way, that the styles are by default injected to head tag in browser. If you

Reviewer 2: ```suggestion
HDS React components use CSS-in-JS in a way where the styles are by default injected into the head tag in the browser
```

use server-side rendering and do not include the critical styles on the server, this leads to flashes of unstyled

> content when user lands to the page.

Reviewer 2: **the** user lands ~~to~~ **on**

Author: - [ ] Fix articles

> To include critical styles in server, HDS exposes a tool for extracting the used critical styles of HDS components.

Reviewer: "**on the** server". Google says that it's ok to use either in or on prepositions with server but there seems to be some consensus among programmers that on is a bit more logical.

Author: - [ ] <b>on the</b> server

> This should come handy in multiple ways:

Reviewer: "--come **in** handy in--"

Author: - [ ] add <b>in</b>

- Automatically extracts styles based on used HTML.
- If you add more HDS components later, you do not need to remember to add their styles separately.
- If you happen to remove HDS components, you do not need to remember to remove their respective styles.

Let's go through a simple example of the usage of the tool:

```js
import { getCriticalHdsRules, hdsStyles } from "hds-react";

const criticalHdsRules = await getCriticalHdsRules(bodyHTML, hdsStyles);

const finalHTML =
<html>
  <head>
    ...
    <style data-used-styles dangerouslySetInnerHTML={{ __html: criticalHdsRules }} />
  </head>
  <body>
    ...
  </body>
<html>
```

First, we import the tool that is named `getCriticalHdsRules`. To be able to call it, we need two things:

> 1. The initially to be rendered HTML body code as a string.

Reviewer: I think one should use lines in this, so "to-be-rendered". I'm not 100% sure though.

Author: - [ ] to be rendered -> to-be-rendered

> 2. hdsStyles - a variable holding all the styles of the HDS react components.

Reviewer 2: HDS react components **as a string**

Author: - [ ] Add <b>as a string</b>

Calling it returns a string containing the critical CSS styles. The rest is easy, we set those styles into
a style tag, that we will then inject into the finally rendered HTML document. It can be wise to cache the result of

> `getCriticalHdsRules` based on the function parameters, to improve performance.

Reviewer: This might nit picking and completely up to preference but I feel like this reads smoother: "--based on the function parameters in order to improve performance." But go ahead and use whatever feels more natural to you.

Author: - [ ] based on the function parameters, to improve performance. -> based on the function parameters in order to improve performance.
